src/rust/ensogl/src/display/scene.rs

```diff
@@ -120,23 +129,84 @@ pub enum Target {
 }
 
 impl Target {
-    fn to_internal(&self) -> Vector4<u32> {
+
+    /// Encode two u32 values into three u8 values.
+    ///
+    /// This is the same encoding that is used in the `fragment_runner`. This encoding is lossy and
+    /// can only encode values up to 12^2=4096 each
+    ///
+    /// We use 12 bits from each value and pack them into the 3 output bytes like described in the
+    /// following diagram.
+    ///
+    ///
+    ///  Input
+    ///
+    ///    value1 (v1) as bytes               value2 (v2) as bytes
+    ///   +-----+-----+-----+-----+           +-----+-----+-----+-----+
+    ///   |     |     |     |     |           |     |     |     |     |
+    ///   +-----+-----+-----+-----+           +-----+-----+-----+-----+
+    /// 32    24    16     8      0         32    24    16     8      0   <- Bit index
+    ///
+    ///
+    /// Output
+    ///
+    /// byte1            byte2                     byte3
+    /// +-----------+    +-------------------+     +-----------+
+    /// | v1[12..4] |    | v1[4..0] v2[4..0] |     | v2[12..4] |
+    /// +-----------+    +-------------------+     +-----------+
+    ///
+    fn encode(value1:u32, value2:u32) -> DecodingResult {
+        let chunk1 = (value1 >> 4u32) & 0x00FFu32;
+        let chunk2 = (value1 & 0x000Fu32) << 4u32;
+        let chunk2 = chunk2 | ((value2 & 0x0F00u32) >> 8u32);
+        let chunk3 = value2 & 0x00FFu32;
+
+        if value1 > 2u32.pow(12) ||value2 > 2u32.pow(12) {
+            DecodingResult::Truncated(chunk1 as u8, chunk2 as u8, chunk3 as u8)
+        }else{
+            DecodingResult::Ok(chunk1 as u8, chunk2 as u8, chunk3 as u8)
+        }
+    }
+
+    /// Decode the symbol_id and instance_id that was encoded in the `fragment_runner`.
+    ///
+    /// See the `encode` method for more information on the encoding.
+    fn decode(chunk1:u32, chunk2:u32, chunk3:u32) -> (u32, u32) {
+        let value1 = (chunk1 << 4) + (chunk2 >> 4);
+        let value2 = chunk3 + ((chunk2 & 0x000F) << 8);
+        (value1, value2)
+    }
+
+    fn to_internal(&self, logger:Option<&Logger>) -> Vector4<u32> {
```

Comment:
Why logger is in option? Just lets assume you need to pass it always, the code will be simpler. Moreover, I don't believe  that allowing not passing logger and dropping warnings then is a good idea.